### PR TITLE
add fp16 support for native conv and deconv on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99498
* __->__ #99497

### Testing

Native conv vs. mkldnn conv on SPR (with avx512_fp16 support)

Single core:

Input | Naïve impl   / us | oneDNN /   us | Speed up
-- | -- | -- | --
IC:   64, OC: 256, kernel: 1, stride: 1, N: 256, H: 56, W: 56, G: 1, pad: 0 | 34676789 | 524199.8 | 66.15185
IC:   128, OC: 512, kernel: 1, stride: 1, N: 256, H: 28, W: 28, G: 1, pad: 0 | 33454125 | 349844.4 | 95.62573
IC: 256, OC: 256, kernel: 3, stride: 1,   N: 1, H: 16, W: 16, G: 1, pad: 0 | 317650.1 | 2317.677 | 137.0554
IC: 128, OC: 256, kernel: 3, stride: 1,   N: 1, L: 64 | 15334.68 | 167.264 | 91.67952

56 cores:
Input | Naïve impl   / us | oneDNN /   us | Speed up
-- | -- | -- | --
IC:   64, OC: 256, kernel: 1, stride: 1, N: 256, H: 56, W: 56, G: 1, pad: 0 | 1032064 | 11073.58 | 93.20061
IC:   128, OC: 512, kernel: 1, stride: 1, N: 256, H: 28, W: 28, G: 1, pad: 0 | 1000097 | 16371.19 | 61.08883
IC:   256, OC: 1024, kernel: 1, stride: 1, N: 256, H: 14, W: 14, G: 1, pad: 0 | 981813.4 | 9008.908 | 108.9825
IC: 1024, OC: 256, kernel: 1, stride: 1,   N: 256, H: 14, W: 14, G: 1, pad: 0 | 1082606 | 10150.47 | 106.6558
IC: 256, OC: 256, kernel: 3, stride: 1,   N: 1, H: 16, W: 16, G: 1, pad: 0 | 319980.6 | 181.598 | 1762.027

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -10496,6 +10503,11 @@ def get_samples():
             elif op.name in self.FP16_LOW_PRECISION_LIST and dtype == torch.float16:
                 atol = 1e-2
                 rtol = 1e-2
+            elif op.name in ['nn.functional.conv_transpose1d',
+                             'nn.functional.conv_transpose2d',
+                             'nn.functional.conv_transpose3d'] and dtype == torch.float16:
+                atol = 5e-2
+                rtol = 5e-2
```
Comment (jgong5) [inline test/test_mps.py:10510]: This has higher error than the normal float16?

File: test/test_mps.py [hunk 2]
```diff
@@ -162,6 +162,13 @@ def mps_ops_grad_modifier(ops):
 
     XPASSLIST_GRAD = {
         'nn.functional.pairwise_distance': [torch.float16],
+        # failed assertion `destination datatype must be fp32'
```
Comment (jgong5) [inline test/test_mps.py:165]: what's the problem here?

Comment (CaoE) [reply test/test_mps.py:165]: The error is : 
```
test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_conv1d_cpu_float16 /AppleInternal/Library/BuildRoots/c651a45f-806e-11ed-a221-7ef33c48bc85/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSNDArray/Kernels/MPSNDArrayConvolutionA14.mm:3976: failed assertion `destination datatype must be fp32'
```
mps does not support fp16 conv.

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@cpuhrsch Could you please review this PR ? Thanks.

**Comment by cpuhrsch** [comment]
@CaoE - Do you think it'd be good for me to wait for https://github.com/pytorch/pytorch/pull/99496 to run green and merge first?

**Comment by CaoE** [comment]
@cpuhrsch Yes. Let's wait for https://github.com/pytorch/pytorch/pull/99496 to merge

**Comment by CaoE** [comment]
@cpuhrsch https://github.com/pytorch/pytorch/pull/99496 has been merged. Could you please review this PR ? Thanks !

**Comment by cpuhrsch** [comment]
@CaoE - There are three failing checks. Are these relevant?

**Comment by CaoE** [comment]
@cpuhrsch No, they are unrelated.

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**